src/Microsoft.AspNetCore.TestHost/WebSocketClient.cs

```diff
@@ -53,73 +53,20 @@ public Action<HttpRequest> ConfigureRequest
-
-            if (ConfigureRequest != null)
-            {
-                ConfigureRequest(state.Context.HttpContext.Request);
-            }
-
-            // Async offload, don't let the test code block the caller.
-            var offload = Task.Factory.StartNew(async () =>
+            WebSocketFeature webSocketFeature = null;
+            var contextBuilder = new HttpContextBuilder(_application);
+            contextBuilder.Configure(context =>
             {
-                try
-                {
-                    await _application.ProcessRequestAsync(state.Context);
-                    state.PipelineComplete();
-                    state.ServerCleanup(exception: null);
-                }
-                catch (Exception ex)
-                {
-                    state.PipelineFailed(ex);
-                    state.ServerCleanup(ex);
-                }
-                finally
-                {
-                    state.Dispose();
-                }
-            });
-
-            return await state.WebSocketTask;
-        }
-
-        private class RequestState : IDisposable, IHttpWebSocketFeature
-        {
-            private readonly IHttpApplication<Context> _application;
-            private TaskCompletionSource<WebSocket> _clientWebSocketTcs;
-            private CancellationTokenRegistration _cancellationTokenRegistration;
-            private WebSocket _serverWebSocket;
-
-            public Context Context { get; private set; }
-            public Task<WebSocket> WebSocketTask { get { return _clientWebSocketTcs.Task; } }
-
-            public RequestState(Uri uri, PathString pathBase, CancellationToken cancellationToken, IHttpApplication<Context> application)
-            {
-                _clientWebSocketTcs = new TaskCompletionSource<WebSocket>();
-                _cancellationTokenRegistration = cancellationToken.Register(
-                    () => _clientWebSocketTcs.TrySetCanceled(cancellationToken));
-                _application = application;
-
-                // HttpContext
-                var contextFeatures = new FeatureCollection();
-                contextFeatures.Set<IHttpRequestFeature>(new RequestFeature());
-                contextFeatures.Set<IHttpResponseFeature>(new ResponseFeature());
-                Context = _application.CreateContext(contextFeatures);
-                var httpContext = Context.HttpContext;
-
-                // Request
-                var request = httpContext.Request;
-                request.Protocol = "HTTP/1.1";
+                var request = context.Request;
                 var scheme = uri.Scheme;
                 scheme = (scheme == "ws") ? "http" : scheme;
                 scheme = (scheme == "wss") ? "https" : scheme;
                 request.Scheme = scheme;
-                request.Method = "GET";
+                request.Method = HttpMethods.Get;
                 var fullPath = PathString.FromUriComponent(uri);
```

Comment:
nit: let's use the same logic as `ClientHandler` https://github.com/aspnet/Hosting/pull/1248/files#diff-22356e5b7de1c9d8b3046bd1c6418043R81